File: cracking_the_coding_interview/stack_queue/stack.php
<?php

class Stack {

  public $stack ;
  public $min = 0;

  function __construct(){
    $this->stack = [];
  }

  public function pop(){
    $poped = array_pop($this->stack);
    echo $poped." is popped from stack";
    if($this->min  $data){
      $this->min = $data;
    }
    return $poped;
  }

  public function push($data){
    echo $data." pushed to stack";
    array_push($this->stack,$data);
    if($this->min > $data){
      $this->min = $data;
    }
  }

  public function isEmpty(){
    return (sizeof($this->stack) > 0)? false : true;
  }

  public function peek(){
    return ($this->isEmpty())? -1 : $this->stack[sizeof($this->stack)-1];
  }

  public function min(){
    return $this->min;
  }

}

$stack = new Stack();
$stack->push('1');
$stack->push('2');
$stack->push('3');
$stack->push('4');
echo $stack->peek();
echo $stack->pop();
echo $stack->peek();

File: cracking_the_coding_interview/stack_queue/golang/stack.go
package main

import "fmt"

type Stack struct {
	items []int
}

func (s *Stack) Push(data int) {
	s.items = append(s.items, data)
}

func (s *Stack) Pop() int {
	if len(s.items) == 0 {
		return -1
	}

	result := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]

	return result
}

func (s *Stack) Size() int {
	return len(s.items)
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack) Top() int {

	if len(s.items) == 0 {
		return -1
	}

	return s.items[0]
}

func main() {

	stack := Stack{}

	stack.Push(10)
	stack.Push(13)
	stack.Push(15)
	stack.Push(1)
	stack.Push(2)

	fmt.Println("Pop:", stack.Pop())
	fmt.Println("Pop:", stack.Pop())

	stack.Push(5)

	fmt.Printf("Stack size: %d", stack.Size())

}
